File: src/sokoenginepy/manager/hashed_board_manager/hashed_board_manager.py
import random

from ... import utilities
from ..board_manager import BoardManager
from ..piece import DEFAULT_PIECE_ID


class HashedBoardManager(BoardManager):
    """:class:`.BoardManager` with Zobrist hashing

    Adds Zobrist hashing on top of :class:`.BoardManager` and keeps it up to date
    when pieces are moved.

    Zobrist hash is 64b integer hash derived from positions of all boxes and pushers
    on board.

    When initialized, :class:`.HashedBoardManager` hashes board using positions and
    IDs of boxes and pushers and produces 64b integer hash. After that, whenever
    position of piece changes, this hash is updated. The ``Zobrist`` part means
    hashing is deterministic which then means that undoing movement will return hash
    value to previous one. All this allows for creation of position tables that
    contain many board layouts and can be quickly compared (since we are not
    comparing positions but only hashes of these positions). Being able to quickly
    compare and find current board layout in some big table, speeds up searching
    through game space which is needed for effective solver implementations.

    Boxes with same Sokoban+ ID are treated as equal meaning that if two of these
    boxes switch position, hash will not change. This also means that hash is
    different when Sokoban+ is enabled from the one when it is disabled

    Pushers are all treated equal, meaning that if two pushers switch position, hash
    will not change

    Notes:
        - enabling/disabling Sokoban+ rehashes the board state
        - changing position of pieces only updates existing hash, it doesn't rehash
          whole board. This means that for example undoing box push would "undo" the
          hash value to the one that was before move was preformed
    """

    def __init__(self, variant_board, boxorder=None, goalorder=None):
        super().__init__(variant_board, boxorder, goalorder)
        self._initial_state_hash = None
        self._state_hash = None
        self._pushers_factors = None
        self._boxes_factors = None
        self._solutions_hashes = None
        self._zobrist_rehash()

    def _zobrist_rehash(self):
        """Recalculates Zobrist hash of board position from scratch."""
        distinct_box_plus_ids = set(
            self.box_plus_id(box_id) for box_id in self.boxes_ids
        )

        # We need random numbers only for this number of positions
        board_without_walls_size = self.board.size - len(self.walls_positions)

        random_pool_size = (
            # one random number for each position pusher can occupy
            board_without_walls_size
            # one random number for position each distinct box can occupy
            + len(distinct_box_plus_ids) * board_without_walls_size
            # one random number for initial hash
            + 1
        )

        # generate required set of random numbers
        # state_backup = random.getstate()
        # Seeding to constant always produces same sequence which then ensures
        # that equal board layouts always produce equal hash
        # random.seed(42)
        random_pool = set()
        while len(random_pool) < random_pool_size:
            random_pool.add(random.getrandbits(64))
        random_pool = list(random_pool)
        # random.setstate(state_backup)

        self._initial_state_hash = self._state_hash = random_pool[0]
        random_pool = random_pool[1:]

        # Store position factors for all distinct positions of all distinct
        # boxes
        self._boxes_factors = dict()
        for box_plus_id in distinct_box_plus_ids:
            self._boxes_factors[box_plus_id] = self._insert_wall_zeroes(
                random_pool[:board_without_walls_size]
            )
            random_pool = random_pool[board_without_walls_size:]

        # Store position factors for all distinct pusher positions
        self._pushers_factors = self._insert_wall_zeroes(
            random_pool[:board_without_walls_size]
        )

        for box_id in self.boxes_ids:
            self._state_hash ^= self._boxes_factors[self.box_plus_id(box_id)][
                self.box_position(box_id)
            ]

        for pusher_position in self.pushers_positions.values():
            self._state_hash ^= self._pushers_factors[pusher_position]

    def _insert_wall_zeroes(self, lst):
        src_index = 0

        def choose(position):
            nonlocal src_index
            if position in self.walls_positions:
                return None
            else:
                src_index += 1
                return lst[src_index - 1]

        return [choose(pos) for pos in range(self.board.size)]

    @property
    def state_hash(self):
        """Zobrist hash of current board state."""
        if self._state_hash is None or self._initial_state_hash is None:
            self._zobrist_rehash()
        return self._state_hash

    @property
    def initial_state_hash(self):
        """
        Zobrist hash of initial board state (before any movement happened).
        """
        if self._state_hash is None or self._initial_state_hash is None:
            self._zobrist_rehash()
        return self._initial_state_hash

    def external_state_hash(self, board_state):
        """
        Calculates Zobrist hash of given ``board_state`` as if that ``board_state``
        was applied to initial ``board`` (to board where no movement happened).

        ``board_state`` must meet following requirement:

            len(board_state.boxes_positions) == self.boxes_count
            and len(board_state.boxes_positions) == self.goals_count

        Returns:
            int or None: Value of hash or None if it can't be calculated
        """
        if (
            len(board_state.boxes_positions) != self.boxes_count
            or len(board_state.boxes_positions) != self.goals_count
        ):
            return None

        retv = self.initial_state_hash
        for index, box_position in enumerate(board_state.boxes_positions):
            retv ^= self._boxes_factors[self.box_plus_id(DEFAULT_PIECE_ID + index)][
                box_position
            ]

        for pusher_position in board_state.pushers_positions:
            retv ^= self._pushers_factors[pusher_position]

        board_state.zobrist_hash = retv

        return retv

    def _box_moved(self, old_position, to_new_position):
        if old_position != to_new_position:
            box_plus_id = self.box_plus_id(self.box_id_on(to_new_position))
            self._state_hash ^= self._boxes_factors[box_plus_id][old_position]
            self._state_hash ^= self._boxes_factors[box_plus_id][to_new_position]

    def _pusher_moved(self, old_position, to_new_position):
        if old_position != to_new_position:
            self._state_hash ^= self._pushers_factors[old_position]
            self._state_hash ^= self._pushers_factors[to_new_position]

    @BoardManager.boxorder.setter
    def boxorder(self, rv):
        old_plus_enabled = self.is_sokoban_plus_enabled
        BoardManager.boxorder.fset(self, rv)
        if self.is_sokoban_plus_enabled != old_plus_enabled:
            self._solutions_hashes = None
            self._zobrist_rehash()

    @BoardManager.goalorder.setter
    def goalorder(self, rv):
        old_plus_enabled = self.is_sokoban_plus_enabled
        BoardManager.goalorder.fset(self, rv)
        if self.is_sokoban_plus_enabled != old_plus_enabled:
            self._solutions_hashes = None
            self._zobrist_rehash()

    def enable_sokoban_plus(self):
        old_plus_enabled = self.is_sokoban_plus_enabled
        super().enable_sokoban_plus()
        if self.is_sokoban_plus_enabled != old_plus_enabled:
            self._solutions_hashes = None
            self._zobrist_rehash()

    def disable_sokoban_plus(self):
        old_plus_enabled = self.is_sokoban_plus_enabled
        super().disable_sokoban_plus()
        if self.is_sokoban_plus_enabled != old_plus_enabled:
            self._solutions_hashes = None
            self._zobrist_rehash()

    @property
    def is_solved(self):
        return self.state_hash in self.solutions_hashes

    @property
    def solutions_hashes(self):
        if not self._solutions_hashes:
            self._solutions_hashes = set(
                h
                for h in (
                    self.external_state_hash(solution) for solution in self.solutions()
                )
                if h
            )
        return self._solutions_hashes

    @property
    def state(self):
        retv = super().state
        retv.zobrist_hash = self.state_hash
        return retv

File: src/libsokoengine/game/hashed_board_manager.cpp
#include "hashed_board_manager.hpp"

#include "board_cell.hpp"
#include "board_graph.hpp"
#include "board_state.hpp"
#include "sokoban_plus.hpp"

#include <chrono>
#include <random>
#include <unordered_set>

#include <boost/algorithm/string.hpp>
#include <boost/lexical_cast.hpp>

using namespace std;

namespace sokoengine {
namespace game {

using io::Strings;

class LIBSOKOENGINE_LOCAL HashedBoardManager::PIMPL {
public:
  bool m_hash_invalidated = true;
  zobrist_key_t m_initial_state_hash = 0;
  zobrist_key_t m_state_hash = 0;

  typedef vector<zobrist_key_t> hash_vector_t;

  map<piece_id_t, hash_vector_t> m_boxes_factors;
  hash_vector_t m_pushers_factors;
  HashedBoardManager::solutions_hashes_t m_solutions_hashes;

  PIMPL() {}
  PIMPL(PIMPL &&rv) = default;
  PIMPL &operator=(PIMPL &&rv) = default;

  static std::string to_str(const hash_vector_t &v, int add_indent = 0) {
    string indent(add_indent, ' ');

    auto converter = [&]() {
      Strings retv;

      Strings tmp;
      for (auto id : v) {
        tmp.push_back(boost::lexical_cast<string>(id));
        if (tmp.size() == 5) {
          retv.push_back(indent + "    " + boost::join(tmp, ", "));
          tmp.clear();
        }
      }
      if (!tmp.empty()) { retv.push_back(indent + "    " + boost::join(tmp, ", ")); }

      return retv;
    };

    return string("[\n") + boost::join(converter(), ",\n") + "\n" + indent + "]";
  }

  static std::string to_str(const map<piece_id_t, hash_vector_t> &m,
                            int add_indent = 0) {
    string indent(add_indent, ' ');
    auto converter = [&]() {
      Strings retv;
      for (auto p : m) {
        retv.push_back(indent + "    " + boost::lexical_cast<string>(p.first) + ": " +
                       to_str(p.second, 4 + add_indent));
      }
      return retv;
    };

    return string("{\n") + boost::join(converter(), ",\n") + "\n" + indent + "}";
  }

  hash_vector_t unique_random_keys(size_t of_size) const {
    unordered_set<zobrist_key_t> random_pool_set;
    mt19937_64 generator;

    generator.seed(chrono::system_clock::now().time_since_epoch().count());
    uniform_int_distribution<zobrist_key_t> distribution;
    auto random_key = bind(distribution, generator);

    while (random_pool_set.size() != of_size) {
      random_pool_set.insert(random_key());
    }

    return hash_vector_t(random_pool_set.begin(), random_pool_set.end());
  }

  void zobrist_rehash(const HashedBoardManager &parent) {
    if (!m_hash_invalidated) return;
    m_hash_invalidated = false;

    set<piece_id_t> distinct_box_plus_ids;
    for (auto box_id : parent.boxes_ids()) {
      distinct_box_plus_ids.insert(parent.box_plus_id(box_id));
    }

    board_size_t board_without_walls_size =
      parent.board().vertices_count() - parent.walls_positions().size();

    size_t random_pool_size = board_without_walls_size +
                              distinct_box_plus_ids.size() * board_without_walls_size +
                              1;

    hash_vector_t random_pool = unique_random_keys(random_pool_size);

    size_t index = 0;
    m_initial_state_hash = m_state_hash = random_pool[index++];

    m_boxes_factors = map<piece_id_t, hash_vector_t>();
    for (auto box_plus_id : distinct_box_plus_ids) {
      m_boxes_factors[box_plus_id] = hash_vector_t();
      for (board_size_t i = 0; i < parent.board().vertices_count(); i++) {
        if (std::find(parent.walls_positions().begin(), parent.walls_positions().end(),
                      i) != parent.walls_positions().end())
          m_boxes_factors[box_plus_id].push_back(0);
        else
          m_boxes_factors[box_plus_id].push_back(random_pool[index++]);
      }
    }
    for (board_size_t i = 0; i < parent.board().vertices_count(); i++) {
      if (std::find(parent.walls_positions().begin(), parent.walls_positions().end(),
                    i) != parent.walls_positions().end())
        m_pushers_factors.push_back(0);
      else
        m_pushers_factors.push_back(random_pool[index++]);
    }

    // Hash from boxes positions
    for (auto box_id : parent.boxes_ids()) {
      m_state_hash ^=
        m_boxes_factors[parent.box_plus_id(box_id)][parent.box_position(box_id)];
    }

    for (auto pusher_id : parent.pushers_ids()) {
      m_state_hash ^= m_pushers_factors[parent.pusher_position(pusher_id)];
    }
  }
}; // HashedBoardManager::PIMPL

HashedBoardManager::HashedBoardManager(BoardGraph &board, const string &boxorder,
                                       const string &goalorder)
  : BoardManager(board, boxorder, goalorder), m_impl(std::make_unique<PIMPL>()) {}

HashedBoardManager::HashedBoardManager(HashedBoardManager &&) = default;

HashedBoardManager &HashedBoardManager::operator=(HashedBoardManager &&) = default;

HashedBoardManager::~HashedBoardManager() = default;

bool HashedBoardManager::operator==(const HashedBoardManager &rv) const {
  return m_impl->m_state_hash == rv.m_impl->m_state_hash;
}

bool HashedBoardManager::operator!=(const HashedBoardManager &rv) const {
  return !(*this == rv);
}

zobrist_key_t HashedBoardManager::state_hash() const {
  const_cast<HashedBoardManager *>(this)->m_impl->zobrist_rehash(*this);
  return m_impl->m_state_hash;
}

zobrist_key_t HashedBoardManager::initial_state_hash() const {
  const_cast<HashedBoardManager *>(this)->m_impl->zobrist_rehash(*this);
  return m_impl->m_initial_state_hash;
}

zobrist_key_t HashedBoardManager::external_state_hash(BoardState &board_state) const {
  if (board_state.boxes_positions().size() != boxes_count() ||
      board_state.boxes_positions().size() != goals_count())
    return 0;

  const_cast<HashedBoardManager *>(this)->m_impl->zobrist_rehash(*this);

  zobrist_key_t retv = m_impl->m_initial_state_hash;

  piece_id_t index = 0;
  for (auto box_position : board_state.boxes_positions()) {
    retv ^=
      m_impl
        ->m_boxes_factors[box_plus_id(Config::DEFAULT_PIECE_ID + index)][box_position];
    index++;
  }

  for (auto pusher_position : board_state.pushers_positions())
    retv ^= m_impl->m_pushers_factors[pusher_position];

  board_state.zobrist_hash() = retv;

  return retv;
}

void HashedBoardManager::pusher_moved(position_t old_position,
                                      position_t to_new_position) {
  if (old_position != to_new_position) {
    const_cast<HashedBoardManager *>(this)->m_impl->zobrist_rehash(*this);
    m_impl->m_state_hash ^= m_impl->m_pushers_factors[old_position];
    m_impl->m_state_hash ^= m_impl->m_pushers_factors[to_new_position];
  }
}

void HashedBoardManager::box_moved(position_t old_position,
                                   position_t to_new_position) {
  if (old_position != to_new_position) {
    const_cast<HashedBoardManager *>(this)->m_impl->zobrist_rehash(*this);
    auto b_plus_id = box_plus_id(box_id_on(to_new_position));
    m_impl->m_state_hash ^= m_impl->m_boxes_factors[b_plus_id][old_position];
    m_impl->m_state_hash ^= m_impl->m_boxes_factors[b_plus_id][to_new_position];
  }
}

void HashedBoardManager::set_boxorder(const std::string &rv) {
  bool old_plus_enabled = is_sokoban_plus_enabled();
  BoardManager::set_boxorder(rv);
  if (is_sokoban_plus_enabled() != old_plus_enabled) {
    m_impl->m_solutions_hashes.clear();
    m_impl->m_hash_invalidated = true;
  }
}

void HashedBoardManager::set_goalorder(const std::string &rv) {
  bool old_plus_enabled = is_sokoban_plus_enabled();
  BoardManager::set_goalorder(rv);
  if (is_sokoban_plus_enabled() != old_plus_enabled) {
    m_impl->m_solutions_hashes.clear();
    m_impl->m_hash_invalidated = true;
  }
}

void HashedBoardManager::enable_sokoban_plus() {
  if (!is_sokoban_plus_enabled()) {
    BoardManager::enable_sokoban_plus();
    m_impl->m_solutions_hashes.clear();
    m_impl->m_hash_invalidated = true;
  }
}

void HashedBoardManager::disable_sokoban_plus() {
  if (is_sokoban_plus_enabled()) {
    BoardManager::disable_sokoban_plus();
    m_impl->m_solutions_hashes.clear();
    m_impl->m_hash_invalidated = true;
  }
}

bool HashedBoardManager::is_solved() const {
  return solutions_hashes().count(state_hash()) > 0;
}

const HashedBoardManager::solutions_hashes_t &
HashedBoardManager::solutions_hashes() const {
  if (m_impl->m_solutions_hashes.empty()) {
    // regenerate hashes
    auto slns = solutions();
    for (auto solution : slns) {
      const_cast<HashedBoardManager *>(this)->m_impl->m_solutions_hashes.insert(
        external_state_hash(solution));
    }
  }

  return m_impl->m_solutions_hashes;
}

string HashedBoardManager::str() const {
  string retv = BoardManager::str();
  boost::replace_all(retv, "<BoardManager pushers:", "<HashedBoardManager pushers:");
  boost::replace_all(retv, "              boxes:", "                    boxes:");
  boost::replace_all(retv, "              goals:", "                    goals:");
  boost::replace_all(retv, "              walls:", "                    walls:");
  boost::replace_all(retv, "              boxorder:", "                    boxorder:");
  boost::replace_all(retv,
                     "              goalorder:", "                    goalorder:");
  boost::replace_all(retv, "              board:", "                    board:");
  return retv;
}

BoardState HashedBoardManager::state() const {
  auto retv = BoardManager::state();
  retv.zobrist_hash() = state_hash();
  return retv;
}

} // namespace game
} // namespace sokoengine
